### PR TITLE
Creat todos component

### PR DESCRIPTION
**Pull request by at-longnguyen**
Initial commit for todos component

### CODE REVIEW

File: src/app/features/todos/Todo.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+export class Todo {
```
Comment (at-vinguyen) [inline src/app/features/todos/Todo.ts]: Rank B
Naming convention, the file name should be `todo.model.ts`

File: src/app/features/todos/Todo.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+    id: number;
+    content: string;
+    check: boolean;
+    constructor(content = ''){
+      this.id = new Date().getTime();
+      this.content = content;
+      this.check = false;
+    }
+  }
```
Comment (at-vinguyen) [inline src/app/features/todos/Todo.ts]: Rank B
Adding new line & keeping indent

File: src/app/features/todos/item-left.pipe.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { Pipe, PipeTransform } from '@angular/core';
```
Comment (at-vinguyen) [inline src/app/features/todos/item-left.pipe.ts]: Rank C
Don't need to implement this pipe

File: src/app/features/feature.module.ts [hunk 4]
```diff
@@ -23,7 +29,13 @@ import { SharedModule } from 'app/shared/module/shared.module';
...
   ],
   declarations: [
-    FeatureComponent
+    FeatureComponent,
+    TodosComponent,
+    TodosHeaderComponent,
+    TodosFooterComponent,
+    TodosMainComponent,
+    FilterTodoPipe,
+    ItemLeftPipe
```
Comment (at-vinguyen) [inline src/app/features/feature.module.ts]: Rank C
Don't need to implement this pipe

File: src/app/features/todos/filter-todo.pipe.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+@Pipe({
+  name: 'filterTodo'
+})
+export class FilterTodoPipe implements PipeTransform {
+
+  transform(allTodos: Todo[], action: String): Todo[] {
+    let displayTodos: Todo[];
+    switch(action) {
+      case '1': displayTodos = [...allTodos]; break;
```
Comment (at-vinguyen) [inline src/app/features/todos/filter-todo.pipe.ts]: Rank B
Using string instead. Ex: `all` `completed` `active`

File: src/app/features/todos/filter-todo.pipe.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import { Pipe, PipeTransform } from '@angular/core';
+import { Todo } from './Todo';
+
+@Pipe({
+  name: 'filterTodo'
+})
+export class FilterTodoPipe implements PipeTransform {
+
+  transform(allTodos: Todo[], action: String): Todo[] {
+    let displayTodos: Todo[];
```
Comment (at-vinguyen) [inline src/app/features/todos/filter-todo.pipe.ts]: Rank C
I think we don't need to define this variable

File: src/app/features/todos/todos-footer/todos-footer.component.html [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+  <span class="todos-footer-col actor">
+    <button id='4' class="guardian" (click)='handleFilter($event)'>Clear completed</button>
+  </span>
+  <ul class="todos-footer-col filter">
+    <li><button id='1' class="_2VhED" (click)='handleFilter($event)'>All</button></li>
+    <li><button id='2' class="_2VhED" (click)='handleFilter($event)'>Active</button></li>
+    <li><button id='3' class="_2VhED" (click)='handleFilter($event)'>Completed</button></li>
+  </ul>
+  <span class="todos-footer-col counter"><strong>{{ itemLeft }}</strong> items left</span>
+</footer>
```
Comment (at-vinguyen) [inline src/app/features/todos/todos-footer/todos-footer.component.html]: Rank C
Adding new line

File: src/app/features/todos/todos-footer/todos-footer.component.html [hunk 8]
```diff
@@ -0,0 +1,11 @@
+<footer class="todos-footer">
+  <span class="todos-footer-col actor">
+    <button id='4' class="guardian" (click)='handleFilter($event)'>Clear completed</button>
```
Comment (at-vinguyen) [inline src/app/features/todos/todos-footer/todos-footer.component.html]: Rank B
why id ???
Use " instead

File: src/app/features/todos/todos-header/todos-header.component.html [hunk 9]
```diff
@@ -0,0 +1,3 @@
+<header class="todos-header">
+  <input class="_1LCq9" type="text" [(ngModel)]='newTodo.content' (keyup.enter)='addNewTodo()' placeholder="What needs to be done?" value="">
```
Comment (at-vinguyen) [inline src/app/features/todos/todos-header/todos-header.component.html]: Rank B
Fix convention

File: src/app/features/todos/todos-header/todos-header.component.html [hunk 10]
```diff
@@ -0,0 +1,3 @@
...
+</header>
```
Comment (at-vinguyen) [inline src/app/features/todos/todos-header/todos-header.component.html]: Rank C
Adding new line

File: src/app/features/todos/todos-header/todos-header.component.ts [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+  ngOnInit() {
+    this.addNewTodo();
+  }
+
+  async addNewTodo() {
+    this.newTodo.id = new Date().getTime();
+    await this.todo.emit(this.newTodo);
+    this.newTodo = new Todo();
+    console.log(this.newTodo);
```
Comment (at-vinguyen) [inline src/app/features/todos/todos-header/todos-header.component.ts]: Rank C
Removing console.log

File: src/app/features/todos/todos-header/todos-header.component.ts [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+  constructor() {
+    this.newTodo = new Todo();
+   }
+
+  ngOnInit() {
+    this.addNewTodo();
+  }
+
+  async addNewTodo() {
```
Comment (at-vinguyen) [inline src/app/features/todos/todos-header/todos-header.component.ts]: Rank C
Why `async` `await`?

File: src/app/features/todos/todos-main/todos-main.component.html [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+                class="item-check" 
+                [checked]="todo.check"
+                name='{{ todo.id }}' 
+                id='{{ todo.id }}' 
+                (click)='checkTodo(todo)'>
+        <label>{{todo.content}}</label>
+        <button class="_6oXTL" (click)='deleteTodo(todo)'></button></div>
+    </li>
+  </ul>
+</section>
```
Comment (at-vinguyen) [inline src/app/features/todos/todos-main/todos-main.component.html]: Rank C
Adding new line

File: src/app/features/todos/todos-main/todos-main.component.html [hunk 14]
```diff
@@ -0,0 +1,15 @@
+<section class="todos-main">
+  <ul class="todo-list" *ngFor='let todo of allTodos'>
+    <li class="todo-item">
+      <div class="R2I1d">
+        <input  type="checkbox" 
+                class="item-check" 
+                [checked]="todo.check"
+                name='{{ todo.id }}'
```
Comment (at-vinguyen) [inline src/app/features/todos/todos-main/todos-main.component.html]: Rank C
HTML/Template Convention

File: src/app/features/todos/todos.component.html [hunk 15]
```diff
@@ -0,0 +1,3 @@
+<app-todos-header (todo)='addNewTodo($event)'></app-todos-header>
+<app-todos-main [allTodos]='allTodos | filterTodo:action' (flag)='check()' (deletedTodo)='deleteTodo($event)'></app-todos-main>
+<app-todos-footer (action)='handleAction($event)' [itemLeft]='allTodos | itemLeft:flag'></app-todos-footer>
```
Comment (at-vinguyen) [inline src/app/features/todos/todos.component.html]: Rank C
Adding new line

File: src/app/features/todos/todos.component.html [hunk 16]
```diff
@@ -0,0 +1,3 @@
+<app-todos-header (todo)='addNewTodo($event)'></app-todos-header>
```
Comment (at-vinguyen) [inline src/app/features/todos/todos.component.html]: Rank C
HTML/Template convention

File: src/app/shared/partial/add-to-homescreen/add-to-homescreen.component.html [hunk 17]
```diff
@@ -12,4 +12,4 @@
       </div>
     </div>
   </div>
-</div>
+</div>
```
Comment (at-vinguyen) [inline src/app/shared/partial/add-to-homescreen/add-to-homescreen.component.html]: Rank C
Adding new line

File: src/app/features/todos/todos-footer/todos-footer.component.html (hunk 8 above)
Comment (at-vinguyen) [reply src/app/features/todos/todos-footer/todos-footer.component.html]: This button makes the todos changed directly, not filter.

File: src/app/features/todos/filter-todo.pipe.ts [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+  name: 'filterTodo'
+})
+export class FilterTodoPipe implements PipeTransform {
+
+  transform(allTodos: Todo[], action: String): Todo[] {
+    switch(action) {
+      case 'all': return [...allTodos]; 
+      case 'active': return allTodos.filter(event => !event.check); 
+      case 'completed': return allTodos.filter(event => event.check);
+      default: return [];
```
Comment (at-vinguyen) [inline src/app/features/todos/filter-todo.pipe.ts]: Rank B
Default is `all`

File: src/app/features/todos/todos-footer/todos-footer.component.html [hunk 19]
```diff
@@ -0,0 +1,11 @@
+<footer class="todos-footer">
+  <span class="todos-footer-col actor">
+    <button id="clearCompleted" class="guardian" (click)="handleFilter($event)">Clear completed</button>
```
Comment (at-vinguyen) [inline src/app/features/todos/todos-footer/todos-footer.component.html]: Rank B
You should use HTML attribute `data-*`

File: src/app/features/todos/todos.component.ts [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+@Component({
+  selector: 'app-todos',
+  templateUrl: './todos.component.html',
+  styleUrls: ['./todos.component.scss']
+})
+export class TodosComponent implements OnInit { 
+
+  allTodos: Todo[];
+  action: String;
+  itemLeft: Number;
```
Comment (at-vinguyen) [inline src/app/features/todos/todos.component.ts]: Rank B
Don't need to define this variable

File: src/app/features/todos/todos.component.ts [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+import { Todo } from './Todo.model';
+
+@Component({
+  selector: 'app-todos',
+  templateUrl: './todos.component.html',
+  styleUrls: ['./todos.component.scss']
+})
+export class TodosComponent implements OnInit { 
+
+  allTodos: Todo[];
```
Comment (at-vinguyen) [inline src/app/features/todos/todos.component.ts]: Rank C
The name should be `todos`

File: src/app/features/todos/todos.component.ts [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+  action: String;
+  itemLeft: Number;
+
+  constructor() {}
+
+  ngOnInit() {
+    this.action = 'all';
+    this.getDefaultTodo();
```
Comment (at-vinguyen) [inline src/app/features/todos/todos.component.ts]: Rank B
Don't need to call function, assign directly

File: src/app/features/todos/todos.component.ts [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+    this.check();
+  }
+
+  deleteTodo(todo: Todo) {
+    this.allTodos = this.allTodos.filter(event => event.id !== todo.id);
+    this.check();
+  }
+
+  // Change flag whenever an item is checked
+  check() {
```
Comment (at-vinguyen) [inline src/app/features/todos/todos.component.ts]: Rank B
Don't need to use this function

File: src/app/features/todos/todos.component.html [hunk 24]
```diff
@@ -0,0 +1,3 @@
+<app-todos-header (todo)="addNewTodo($event)"></app-todos-header>
+<app-todos-main [allTodos]="allTodos | filterTodo:action" (flag)="check()" (deletedTodo)="deleteTodo($event)"></app-todos-main>
+<app-todos-footer (action)="handleAction($event)" [itemLeft]="itemLeft"></app-todos-footer>
```
Comment (at-vinguyen) [inline src/app/features/todos/todos.component.html]: Rank B
`[itemLeft]="todo.length"`

File: src/app/features/todos/todos.component.html [hunk 25]
```diff
@@ -0,0 +1,3 @@
+<app-todos-header (todo)="addNewTodo($event)"></app-todos-header>
+<app-todos-main [allTodos]="allTodos | filterTodo:action" (flag)="check()" (deletedTodo)="deleteTodo($event)"></app-todos-main>
```
Comment (at-vinguyen) [inline src/app/features/todos/todos.component.html]: Rank C
what is `(flag)`?

File: src/app/features/todos/filter-todo.pipe.ts [hunk 26]
```diff
@@ -0,0 +1,18 @@
+import { Pipe, PipeTransform } from '@angular/core';
+import { Todo } from './Todo.model';
```
Comment (at-vinguyen) [inline src/app/features/todos/filter-todo.pipe.ts]: Rank C
Naming: `todo.model`

File: src/app/features/todos/filter-todo.pipe.ts [hunk 27]
```diff
@@ -0,0 +1,18 @@
+import { Pipe, PipeTransform } from '@angular/core';
```
Comment (at-vinguyen) [inline src/app/features/todos/filter-todo.pipe.ts]: Rank C
Naming: `todo-filter.pipe.ts`

File: src/app/features/todos/filter-todo.pipe.ts [hunk 28]
```diff
@@ -0,0 +1,18 @@
...
+import { Todo } from './Todo.model';
+
+@Pipe({
+  name: 'filterTodo'
```
Comment (at-vinguyen) [inline src/app/features/todos/filter-todo.pipe.ts]: Rank C
`todoFilter`

File: src/app/features/todos/filter-todo.pipe.ts [hunk 29]
```diff
@@ -0,0 +1,18 @@
...
+})
+export class FilterTodoPipe implements PipeTransform {
```
Comment (at-vinguyen) [inline src/app/features/todos/filter-todo.pipe.ts]: Rank C
`TodoFilterPipe`

File: src/app/features/todos/todos-header/todos-header.component.html [hunk 30]
```diff
@@ -0,0 +1,8 @@
+<header class="todos-header">
+  <input  class="_1LCq9" 
+          type="text"
```
Comment (at-vinguyen) [inline src/app/features/todos/todos-header/todos-header.component.html]: Rank C
type="text" is default

File: src/app/features/todos/todos-main/todos-main.component.ts [hunk 31]
```diff
@@ -0,0 +1,29 @@
...
+  @Input() allTodos: Todo[];
+  @Output() deletedTodo: EventEmitter<Todo> = new EventEmitter();
+  @Output() flag: EventEmitter<Boolean> = new EventEmitter();
+
+  constructor() { }
+
+  ngOnInit() {
+  }
+
+  checkTodo(todo) {
```
Comment (at-vinguyen) [inline src/app/features/todos/todos-main/todos-main.component.ts]: Rank C
Naming by function, it's should be `markAsCompleted`

File: src/app/features/todos/Todo.model.ts [hunk 32]
```diff
@@ -0,0 +1,11 @@
+export class Todo {
+    id: number;
+    content: string;
```
Comment (at-vinguyen) [inline src/app/features/todos/Todo.model.ts]: Rank C
`name`

File: src/app/features/todos/Todo.model.ts [hunk 33]
```diff
@@ -0,0 +1,11 @@
...
+    check: boolean;
```
Comment (at-vinguyen) [inline src/app/features/todos/Todo.model.ts]: Rank C
`isCompleted`

File: src/app/features/todos/Todo.model.ts [hunk 34]
```diff
@@ -0,0 +1,11 @@
...
+    constructor(content = ''){
+      this.id = new Date().getTime();
+      this.content = content;
+      this.check = false;
+    }
+  }
+
```
Comment (at-vinguyen) [inline src/app/features/todos/Todo.model.ts]: Rank C
Adding new line

**Comment by at-vinguyen** [comment]
@at-longnguyen 
Please remove unused files (`scss`)

File: src/app/features/todos/todos-footer/todos-footer.component.html (hunk 19 above)
Comment (at-longnguyen) [reply src/app/features/todos/todos-footer/todos-footer.component.html]: I do not understand what you mean.

**Pull request merged**